app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/OsmQuestController.kt

```diff
@@ -105,19 +106,21 @@ class OsmQuestController internal constructor(
         /** Replace all quests of the given types in the given bounding box with the given quests.
          *  Called on download of a quest type for a bounding box. */
         override fun onReplacedForBBox(bbox: BoundingBox, mapDataWithGeometry: MapDataWithGeometry) {
-            val quests = createQuestsForBBox(bbox, mapDataWithGeometry, allQuestTypes)
+            val quests = createQuestsForBBox(bbox, mapDataWithGeometry, allQuestTypes) // don't put these quests in cache, some could be hidden
             val obsoleteQuestKeys: List<OsmQuestKey>
             synchronized(this) {
-                val previousQuests = db.getAllInBBox(bbox)
+                val previousQuests = db.getAllInBBox(bbox) // can't be done from cache because here we need ALL quests
```

Comment:
Some ideas:
* have some boolean var storing whether a db write in background is currently running, and defer further db access until write is done
* cache *all* quests, also non-visible types and quests hidden by user or by notes (would reduce performance)
* move cache to `VisibleQuestsSource` (no cache for `OsmQuestController.get`)

Reply:
But... if this area is loaded via cache, it means the area might get loaded from db only to be replaced almost immediately.

Reply:
Assuming we would need to cache all quests in `OsmQuestsController`, I'm currently in favor of putting the cache in `VisibleQuestsSource`, because
* loading quests in bbox from db would be faster than with cache in `OsmQuestsController` (because we only need to get visible quests). In some test I did, getting *all* quests in bbox took ca twice as long as getting only quests visible by default.
* the cache would be smaller in size
* the cache will be a little bit faster (no need for querying questTypeRegistry and the 2 questSources/Caches), though I guess the difference might be rather small
* no need for a spatial cache in `NoteController`

The drawbacks would be
* quests / notes can't be persisted asynchronously. But this is pretty fast anyway (*much* faster than persisting elements and geometries).
* no cached access to quests/notes via `Controller.get`. This is really not nice, but could be alleviated by simple LRU caches in both controllers, which are filled with the results of `get(key)` and `get(bbox)`. I would estimate this will not take much memory (mostly the size of the map), as the quests/notes will also be cached by `VisibleQuestsSource`
----
Actually such a simple LRU cache might also be useful for geometries in `MapDataController`:
Usually (visible) quests are loaded (and cached) as first thing when scrolling, which also loads geometry for the related elements.
Later (when enabling an overlay, or when opening a quest which highlights elements), the same geometries will be loaded (and cached) again as part of mapData. So we will end up with many quest geometries loaded twice from db and both being cached, effectively wasting memory used by (estimate) a few thousand duplicate geometries.

Reply:
Hmm, but VisibleQuestsSource is updated from **a lot** of sources, wouldn't it be quite laborous (and thus error prone) to keep it up-to-date?

>  Later (when enabling an overlay, or when opening a quest which highlights elements), the same geometries will be loaded (and cached) again as part of mapData. So we will end up with many quest geometries loaded twice from db and both being cached, effectively wasting memory used by (estimate) a few thousand duplicate geometries.

Yeah, but I don't understand how you would want to solve that. Moving the cache to VisibleQuestSource doesn't, no?

Reply:
You think updating the cache in `VisibleQuestsSource.updateVisibleQuests` is not sufficient?

Reply:
Hmm, no, I guess that would work